core/API/DataTableManipulator/ReportTotalsCalculator.php

```diff
@@ -83,30 +80,76 @@ protected function manipulateDataTable($dataTable)
+        $clone = $firstLevelTable->getEmptyClone($keepFilters = false);
+        foreach ($firstLevelTable->getQueuedFilters() as $queuedFilter) {
+            if (is_array($queuedFilter) && 'ReplaceColumnNames' === $queuedFilter['className']) {
+                $clone->queueFilter($queuedFilter['className'], $queuedFilter['parameters']);
             }
         }
+        $tableMeta = $firstLevelTable->getMetadata(DataTable::COLUMN_AGGREGATION_OPS_METADATA_NAME);
 
+        /** @var DataTable\Row $totalRow */
+        $totalRow = null;
         foreach ($firstLevelTable->getRows() as $row) {
-            $columns = $row->getColumns();
-            foreach ($metricNames as $metricId => $metricName) {
-                $this->sumColumnValueToTotal($columns, $metricId, $metricName);
+            if (!isset($totalRow)) {
+                $columns = $row->getColumns();
+                $columns['label'] = DataTable::LABEL_TOTALS_ROW;
+                $totalRow = new DataTable\Row(array(DataTable\Row::COLUMNS => $columns));
+            } else {
+                $totalRow->sumRow($row, $copyMetadata = false, $tableMeta);
             }
         }
+        $clone->addRow($totalRow);
 
-        $dataTable->setMetadata('totals', $this->totals);
+        if ($this->report
+            && $this->report->getProcessedMetrics()
+            && array_keys($this->report->getProcessedMetrics()) === array('nb_actions_per_visit', 'avg_time_on_site', 'bounce_rate', 'conversion_rate')) {
+            // hack for AllColumns table or default processed metrics
+            $clone->filter('AddColumnsProcessedMetrics', array($deleteRowsWithNoVisit = false));
+        }
+
+        $processor = new DataTablePostProcessor($this->apiModule, $this->apiMethod, $this->request);
+        $processor->applyComputeProcessedMetrics($clone);
+        $clone = $processor->applyQueuedFilters($clone);
+        $clone = $processor->applyMetricsFormatting($clone);
+
+        $totalRow = null;
+        foreach ($clone->getRows() as $row) {
+            /** * @var DataTable\Row $row */
+            if ($row->getColumn('label') === DataTable::LABEL_TOTALS_ROW) {
+                $totalRow = $row;
+                break;
+            }
+        }
+
+        if (!isset($totalRow) && $clone->getRowsCount() === 1) {
+            // if for some reason the processor renamed the totals row,
+            $totalRow = $clone->getFirstRow();
+        }
+
+        if (isset($totalRow)) {
+            $totals = $row->getColumns();
+            unset($totals['label']);
+            $dataTable->setMetadata('totals', $totals);
+
+            if (1 === Common::getRequestVar('keep_totals_row', 0, 'integer', $this->request)) {
+                $row->deleteMetadata(false);
+                $row->setColumn('label', Piwik::translate('General_Totals'));
+                $dataTable->setTotalsRow($row);
+            }
+        }
```

Comment:
I guess this code is what you are referring to in the ticket description? It definitely seems complicated... If the main issue for not doing it the simple way is BC (ie, it might break some reports), maybe we can whitelist which reports support it?

Eg, in the Report class there could be a `supportsTotalsRow()` method that defaults to false, then we could make sure core reports support it even if we have to tweak the API code (eg, changing MultiSites to ignore the totals row). What do you think?

Reply:
Yes, I think now it works quite well though, and might actually even work better than before but not sure if I am missing something.

Reply:
I was thinking about having to clone the datatable and make accomadations for specific filters/processed metrics/etc. What do you think about the idea w/ `supportsTotalsRow()`?

Reply:
Sure could add it, but I think it's not really needed. Can't hurt though maybe to add it and we just set it to `true` by default so a report can opt out. Before it was basically supported on pretty much all reports so be great to still have it on those reports.
It should support pretty much all reports as it is now using 

```
$tableMeta = $firstLevelTable->getMetadata(DataTable::COLUMN_AGGREGATION_OPS_METADATA_NAME);
$totalRow->sumRow($row, $copyMetadata = false, $tableMeta)
``` 
instead of doing the sum in the code directly where I noticed even some problems but forgot which ones. There's some "hack" for some reports that don't add the processed metrics in the report directly `$clone->filter('AddColumnsProcessedMetrics', array($deleteRowsWithNoVisit = false));`  but otherwise it should be fine.